### PR TITLE
Simplify end model scheduler/dataset signatures

### PR DESCRIPTION
**Pull request by bhancock8**
Another stepping stone on the path to fixing up the end model:
- Remove unused add_X/remove_X methods from MultitaskDataset
- Significantly simplify batch schedulers by letting them pass back the dataloader object from which metadata can be pulled; this will also make the trainer more single-task/multi-task agnostic
- Make split and dataset name attributes of the dataset, not the dataloader
- Remove potentially stochastic tests in `test_trainer.py` that checked for quality after training.

**Test plan:**
- Added unit tests for schedulers
- Additional typing coming in #1204

### CODE REVIEW

File: snorkel/end_model/batch_schedulers/sequential_scheduler.py [hunk 1]
```diff
@@ -10,22 +10,8 @@ def __init__(self):
...
-        batch_counts = [len(dataloader) for dataloader in dataloaders]
-        data_loaders = [iter(dataloader) for dataloader in dataloaders]
-        splits = [dataloader.split for dataloader in dataloaders]
-
-        for task_to_label_dict, data_name, batch_count, data_loader, split in zip(
-            task_to_label_dicts, data_names, batch_counts, data_loaders, splits
-        ):
-            for batch in range(batch_count):
-                yield next(data_loader), task_to_label_dict, data_name, split
+        for dataloader in dataloaders:
```
Comment (henryre) [inline snorkel/end_model/batch_schedulers/sequential_scheduler.py]: 😮

File: snorkel/end_model/batch_schedulers/sequential_scheduler.py [hunk 2]
```diff
@@ -10,22 +10,8 @@ def __init__(self):
     def get_batches(self, dataloaders):
         """Return batches in sequential order.
 
-        :param dataloaders: a list of dataloaders
-        :type dataloaders: list
-        :return: A generator of all batches
-        :rtype: genertor
+        TBD
```
Comment (henryre) [inline snorkel/end_model/batch_schedulers/sequential_scheduler.py]: 😠

File: snorkel/end_model/batch_schedulers/shuffled_scheduler.py [hunk 3]
```diff
@@ -12,27 +12,16 @@ def __init__(self):
     def get_batches(self, dataloaders):
         """Return batches in shuffled order.
 
-        :param dataloaders: a list of dataloaders
-        :type dataloaders: list
-        :return: A generator of all batches
-        :rtype: genertor
+        TBD
```
Comment (henryre) [inline snorkel/end_model/batch_schedulers/shuffled_scheduler.py]: 😠

File: snorkel/end_model/batch_schedulers/shuffled_scheduler.py [hunk 4]
```diff
@@ -12,27 +12,16 @@ def __init__(self):
...
         """
+        batch_counts = [len(dl) for dl in dataloaders]
```
Comment (henryre) [inline snorkel/end_model/batch_schedulers/shuffled_scheduler.py]: 😮

File: snorkel/end_model/batch_schedulers/shuffled_scheduler.py [hunk 5]
```diff
@@ -12,27 +12,16 @@ def __init__(self):
...
-        data_names = [dataloader.data_name for dataloader in dataloaders]
-        batch_counts = [len(dataloader) for dataloader in dataloaders]
-        data_loaders = [iter(dataloader) for dataloader in dataloaders]
-        splits = [dataloader.split for dataloader in dataloaders]
-
-        dataloader_indexer = []
+        dataloader_indices = []
         for idx, count in enumerate(batch_counts):
-            dataloader_indexer.extend([idx] * count)
+            dataloader_indices.extend([idx] * count)
```
Comment (henryre) [inline snorkel/end_model/batch_schedulers/shuffled_scheduler.py]: 👍

File: snorkel/end_model/data.py [hunk 6]
```diff
@@ -134,32 +86,19 @@ class MultitaskDataLoader(DataLoader):
...
 
-        for task_name, label_names in task_to_label_dict.items():
-            if not isinstance(label_names, list):
-                label_names = [label_names]
-            unrecognized_labels = set(label_names) - set(dataset.Y_dict.keys())
-            if len(unrecognized_labels) > 0:
-                msg = (
-                    f"Unrecognized Label {unrecognized_labels} of Task {task_name} in "
-                    f"dataset {dataset.name}"
+        for label in self.task_to_label_dict.values():
```
Comment (henryre) [inline snorkel/end_model/data.py]: 😮

File: test/end_model/batch_schedulers/test_schedulers.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        self.assertEqual(data, ["a", "b", "c", "d", "e", "f"])
+
+    def test_shuffled(self):
+        set_seed(123)
+        scheduler = ShuffledScheduler()
+        data = []
+        for (batch, dl) in scheduler.get_batches(dataloaders):
+            X_dict, Y_dict = batch
+            data.extend(X_dict["data"])
+        self.assertNotEqual(data, ["a", "b", "c", "d", "e", "f"])
```
Comment (henryre) [inline test/end_model/batch_schedulers/test_schedulers.py]: 1 / 720 chance is too high for me. Can we do a few more chars?

Comment (henryre) [reply test/end_model/batch_schedulers/test_schedulers.py]: set_seed isn't good enough since the internal implementation might change.

File: test/end_model/test_trainer.py [hunk 8]
```diff
@@ -19,36 +18,24 @@
 class TrainerTest(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
-        cls.trainer_config = {"n_epochs": 3, "progress_bar": False}
+        cls.trainer_config = {"n_epochs": 2, "progress_bar": False}
```
Comment (henryre) [inline test/end_model/test_trainer.py]: can we just put this in the global namespace? it's not modified.
